components/zcash/rs/lib.rs

```diff
@@ -105,3 +212,114 @@ impl OrchardExtendedSpendingKey {
     }
 }
 
+
+impl OrchardBundleValue {
+    fn raw_tx(self: &OrchardBundleValue) -> Vec<u8> {
+        self.0.raw_tx.clone()
+    }
+}
+
+fn create_orchard_builder_internal(
+    orchard_tree_bytes: &[u8],
+    outputs: Vec<OrchardOutput>,
+    random_source: OrchardRandomSource
+) -> Box<OrchardBuilderResult> {
+    use orchard::Anchor;
+    use crate::librustzcash::merkle_tree::read_commitment_tree;
+
+    let anchor = if orchard_tree_bytes.len() > 0 {
+        match read_commitment_tree::<MerkleHashOrchard, _, { orchard::NOTE_COMMITMENT_TREE_DEPTH as u8 }>(
+                &orchard_tree_bytes[..]) {
+            Ok(tree) => Anchor::from(tree.root()),
+            Err(_e) => return Box::new(OrchardBuilderResult::from(Err(Error::from(OrchardBuildError::AnchorMismatch)))),
+        }
+    } else {
+        orchard::Anchor::empty_tree()
+    };
+
+    let mut builder = orchard::builder::Builder::new(
+        orchard::builder::BundleType::DEFAULT,
+        anchor);
+
+    for out in outputs {
+        let _ = match Option::from(orchard::Address::from_raw_address_bytes(&out.addr)) {
+            Some(addr) => {
+                builder.add_output(None, addr,
+                    orchard::value::NoteValue::from_raw(out.value), None)
+            },
+            None => return Box::new(OrchardBuilderResult::from(Err(Error::WrongOutputError)))
+        };
+    }
+
+    Box::new(OrchardBuilderResult::from(match random_source {
+        OrchardRandomSource::OsRng(mut rng) => {
+            builder.build(&mut rng)
+                .map_err(Error::from)
+                .and_then(|builder| {
+                    builder.map(|bundle| OrchardBuilder { unauthorized_bundle: bundle.0, rng: OrchardRandomSource::OsRng(rng) })
+                        .ok_or(Error::BuildError)
+                })
+        },
+        OrchardRandomSource::MockRng(mut rng) => {
```

Comment:
You could avoid the duplication here by implementing a forwarding `RngCore` on the `OrchardRandomSource` wrapper, but that wouldn't be less code so probably not worth it.